arelle/plugin/validate/EDINET/PluginValidationDataExtension.py

```diff
@@ -109,6 +113,60 @@ def shouldValidateUpload(self, val: ValidateXbrl) -> bool:
             return False  # Not a zipfile
         return True
 
+    @lru_cache(1)
+    def getConsolidatedBalanceSheets(self, modelXbrl: ModelXbrl) -> list[BalanceSheet]:
+        return [balanceSheet for balanceSheet in self.getBalanceSheets(modelXbrl) if balanceSheet.consolidated]
+
+    @lru_cache(1)
+    def getBalanceSheets(self, modelXbrl: ModelXbrl) -> list[BalanceSheet]:
+        balanceSheets: list[BalanceSheet] = []
+        deduplicatedFacts = self.getDeduplicatedFacts(modelXbrl)
+        assetsConcept = modelXbrl.qnameConcepts.get(self.assetsIfrsQn)
+        liabilitiesAndEquityConcept = modelXbrl.qnameConcepts.get(self.liabilitiesAndEquityIfrsQn)
+        if assetsConcept is None or liabilitiesAndEquityConcept is None:
+            return balanceSheets
+
+        factsByContextId = defaultdict(list)
+        for fact in deduplicatedFacts:
+            if fact.qname not in (self.assetsIfrsQn, self.liabilitiesAndEquityIfrsQn):
+                continue
+            if fact.contextID is None:
+                continue
+            factsByContextId[fact.contextID].append(fact)
```

Comment:
Should this use `(ContextHashKey(...),UnitHashKey(...))`?  For context, I think there's another requirement that contexts be unique, but perhaps a `getContextKey` helper method that returns `fact.contextID` but has a comment referencing the that error code to justify would be helpful (to avoid copy/paste into other taxonomies where the same might not be true).  For unit, is there a similar requirement that only one unit is used?

Reply:
Units must be unique, but I can't see anything that says only one may be used. I've updated with a helper function that groups the facts by both context ID and unit ID with a note about the GFM rules that allow us to use IDs as keys.